app/models/registration.rb

```diff
@@ -4,13 +4,70 @@ class Registration < ApplicationRecord
   # Professors have extra privileges over assistants.
   enum role: [:student, :grader, :assistant, :professor]
 
+  validates_presence_of :course_id, :user_id
+  validates_inclusion_of :show_in_lists, in: [false, true]
+  
   # A registration is join model between a user and a course.
   belongs_to :user
   belongs_to :course
 
   has_one :term, through: :course
-  belongs_to :section, :primary_key => "crn"
+  has_many :registration_sections, dependent: :destroy
+  has_many :sections, through: :registration_sections
 
+  Section::types.each do |t, value|
+    # This reader is for convenience
+    define_method "#{t}_sections".to_sym do
+      self.sections.where(type: value)
+    end
+  end
+
+  attr_accessor :username
+  attr_accessor :orig_sections
```

Comment:
In the bulk_edit and bulk_update actions, if you edit which section a student is in, then I need to know which section they used to be in so I can update the RegistrationSections table accordingly.  I phrase that using a singular "the section", but the code is phrased in terms of plural "sections", since (a) I'm modelling lectures and labs as sections of a single course, and students are therefore enrolled in multiple sections (currently at most one of each type), and (b) hypothetically a student might one day be enrolled in two sections of the same class (e.g. independent study) and so it's easier to get the plural nature of this correct in advance.

The username attribute is needed instead of user_id, because in the bulk_edit case, I want to be able to paste in a csv of a bunch of student usernames, and they may never yet have logged into BN and created a User record.

There is still a bug in the reg_requests code, which is still in terms of calling Course#add_registration, and that method really ought to go away.  I haven't fixed up the reg_requests stuff yet; there were enough schema-changing things here, and enough code changes (63 files!  jeez, rails) that I wanted to get this landed on its own, first.